GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs

```diff
@@ -21,27 +24,72 @@ public LooseObjectsStep(GVFSContext context, bool requireCacheLock, bool forceRu
         }
 
         public override string Area => nameof(LooseObjectsStep);
+        public int LooseObjectsPutIntoPackFile { get; private set; } = 0;
+        public int MaxLooseObjectsInPack { get; set; } = 100000;
 
         protected override string LastRunTimeFilePath => Path.Combine(this.Context.Enlistment.GitObjectsRoot, "info", LooseObjectsLastRunFileName);
         protected override TimeSpan TimeBetweenRuns => TimeSpan.FromDays(7);
 
         public int CountLooseObjects()
         {
             int count = 0;
-            
+
             foreach (string directoryPath in this.Context.FileSystem.EnumerateDirectories(this.Context.Enlistment.GitObjectsRoot))
             {
                 string directoryName = directoryPath.TrimEnd(Path.DirectorySeparatorChar).Split(Path.DirectorySeparatorChar).Last();
 
                 if (GitObjects.IsLooseObjectsDirectory(directoryName))
-                {                    
+                {
                     count += this.Context.FileSystem.GetFiles(Path.Combine(this.Context.Enlistment.GitObjectsRoot, directoryPath), "*").Count();
                 }
             }
 
             return count;
         }
 
+        public void PackLooseObjects(StreamWriter streamWriter)
+        {
+            // Find loose Objects
+            foreach (DirectoryItemInfo directoryItemInfo in this.Context.FileSystem.ItemsInDirectory(this.Context.Enlistment.GitObjectsRoot))
+            {
+                if (directoryItemInfo.IsDirectory)
+                {
+                    string directoryName = directoryItemInfo.Name;
+
+                    if (GitObjects.IsLooseObjectsDirectory(directoryName))
+                    {
+                        string[] looseObjectFileNamesInDir = this.Context.FileSystem.GetFiles(Path.Combine(directoryItemInfo.FullName), "*");
+
+                        foreach (string file in looseObjectFileNamesInDir)
+                        {
+                            this.LooseObjectsPutIntoPackFile++;
+                            streamWriter.Write(this.GetLooseObjectId(directoryName, file) + "\n");
+
+                            // Stop when we hit the limit.  The next run will pack more files.
+                            if (this.LooseObjectsPutIntoPackFile == this.MaxLooseObjectsInPack)
+                            {
+                                return;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        public string GetLooseObjectId(string directoryName, string filePath)
+        {
+            return directoryName + Path.GetFileName(filePath);
+        }
+
+        public void CreateLooseObjectsPackFile()
+        {
+            GitProcess.Result result = this.RunGitCommand(
+                (process) => process.PackObjects(
+                    "from-loose",

```

Comment:
What will happen if one (or more) of the loose objects is corrupt?  Is there anyway that we can detect that here and delete the corrupt object and try again?

(I would guess this would be a rare event based on how VFS4G downloads loose objects, but it would still be good to know what will happen)

Reply:
the git operation would complain.  I'll update the writer to skip correct objects.  This should never happen but it doesn't hurt to be proactive against errors.

Reply:
It does hurt: we'll slow down a lot if VFS is opening all of these files, too. Better to let Git handle it. Just check to see what `git pack-objects` does on a corrupted object. (Running a test now)

Reply:
Here's what happens:

```

[2018-12-19 13:05:27 -05:00] LooseObjectsStep (Start)
[2018-12-19 13:05:28 -05:00] RunGitCommand (Start) {"Area":"LooseObjectsStep"}
[2018-12-19 13:05:57 -05:00] RunGitCommand (Stop) {"DurationMs":29171}
[2018-12-19 13:05:57 -05:00] RunGitCommand (Start) {"Area":"LooseObjectsStep"}
[2018-12-19 13:06:41 -05:00] Warning {"WarningMessage":"LooseObjectsStep: Git process failed with errors:error: corrupt loose object '2b7d64425d8c559c9424db6cadc7fa2043e2dc23'\nfatal: loose object 2b7d64425d8c559c9424db6cadc7fa2043e2dc23 (stored in C:\\.gvfsCache\\a7414878aa15429f9b63c1a753783c7a\\gitObjects/2b/7d64425d8c559c9424db6cadc7fa2043e2dc23) is corrupt\n"}
[2018-12-19 13:06:41 -05:00] RunGitCommand (Stop) {"DurationMs":44178}
[2018-12-19 13:06:41 -05:00] LooseObjectsStep {"GitObjectsRoot":"C:\\.gvfsCache\\a7414878aa15429f9b63c1a753783c7a\\gitObjects","StartingCount":344526,"EndingCount":244526,"RemovedCount":100000,"LooseObjectsPutIntoPackFile":10000}
[2018-12-19 13:06:42 -05:00] LooseObjectsStep (Stop) {"DurationMs":74395}
```

So, we get a warning, and `git` returns with failure (which makes sense, it can't pack that object!). It does mean that we can track the warning. We have enough data logging to see if users are stuck at a certain value (or only increasing) and investigate. Deleting the corrupt loose object fixes the problem.